### ISSUE TITLE
Linear memory growth leads to Freeze of training

### ISSUE THREAD

**Issue by Max-Ryujin** [rwth-i6/returnn#1450]
# Discription
This issue appeared while testing `feature_args = {"class": "ScfNetwork", "size_tf": 256 // 2, "stride_tf": 10 // 2}` in a CTC setup. 
An identical setup with `feature_args = {"class": "LogMelNetwork", "wave_norm": True, "frame_size": 200, "frame_shift": 80, "fft_size": 256}`
works fine. The training runs for a few epochs while memory usage climbs linearly until the limit is reached. At this point, the step times grow fast until there is nearly no progress at all. In rare cases, the training crashes with an out-of-memory error. I am using the newest version of returnn. This runs in an apptainer.
The setup is `run_scf_baseline_big` in `/u/maximilian.kannen/setups/20230406_feat/recipe/i6_experiments/users/vieting/experiments/switchboard/ctc/feat/experiments.py`
https://gist.github.com/Max-Ryujin/690647f79773d6cd8338c524be039040

# Relavant Logs

## Example Freeze `/u/maximilian.kannen/setups/20230406_feat/alias/experiments/switchboard/ctc/feat/train_nn/conformer_bs5k_scf_baseline_big/log.run.1`
https://gist.github.com/Max-Ryujin/b2da56c72afaf28850dd7384c97a5b2a
## Example OOM Error `/u/maximilian.kannen/setups/20230406_feat/alias/experiments/switchboard/ctc/feat/train_nn/conformer_bs5k_audio_perturbation_scf_conf-wei-oldspecaug-audio_perturbation_speed0.4_0.8_1.2/log.run.1`
https://gist.github.com/Max-Ryujin/c559f169bc27f8f77e1e39af0146282b

**Comment by Max-Ryujin** [comment]
The returnn config:
https://gist.github.com/Max-Ryujin/f7323769568c9803bd19e3eba623c2d1

**Comment by albertz** [comment]
As far as I see, the only thing a bit uncommon is the `fast_bw` loss here, using Sprint (RASR) in a subprocess to compute the FSA. I wonder if that causes the problems.

**Comment by albertz** [comment]
Can you run this in a memory profiler and report the observations?

**Comment by albertz** [comment]
Also, can you add some more details:

- Since when does this happen?
- What did you change after this happens? TensorFlow version? RETURNN version? Something else?

**Comment by albertz** [comment]
> An identical setup with `feature_args = {"class": "LogMelNetwork", "wave_norm": True, "frame_size": 200, "frame_shift": 80, "fft_size": 256}`
> works fine.

Are you sure? Can you try again and post also the log (including watch_memory log) for that?

**Comment by Max-Ryujin** [comment]
I did not change anything. This setup never worked. I tried running this for the first time 3 weeks ago.

**Comment by Max-Ryujin** [comment]
Older log for mel log without watch_memory:
`/u/maximilian.kannen/setups/20230406_feat/alias/experiments/switchboard/ctc/feat/train_nn/conformer_bs5k_mel_baseline/log.run.1`
https://gist.github.com/Max-Ryujin/60d6dcc820ddd0e960e753806ceeb9af

**Comment by albertz** [comment]
As I wrote on Slack: It's very strange that the only change is the feature extraction, as part of the network. So only the TF computation graph changes a bit, nothing else. Maybe the memory leak is actually on TF side? Or maybe on CUDA/CuDNN side? Maybe the different convolution settings in the beginning trigger this somehow. So we should try out newer TF/CUDA/CuDNN.

**Comment by albertz** [comment]
How often do you get hangs vs OOM? How long does it take to get that?

Next time when it hangs, it would be very interesting if you could login to the node and attach with GDB to the running process and then print the backtrace of all running threads (`t apply all bt` in GDB). That might give us a hint whether it is a TF mem leak.

**Comment by Max-Ryujin** [comment]
hangs vs OOM is very hard to judge, because I tested a lot of different amounts of memory and combinations of settings which could lead to different outcomes.

**Comment by Max-Ryujin** [comment]
I just canceled the training that was stuck from last night, but if I start a new one with less memory, it should get stuck in 2-3 hours

**Comment by Max-Ryujin** [comment]
Log mel trainings do not have the same memory leak. The problem is specific to the SCF network architecture.

**Comment by albertz** [comment]
What is the native stacktrace (via GDB)? You should see it hang somewhere inside the native TF lib.

**Comment by albertz** [comment]
> What is the native stacktrace (via GDB)? You should see it hang somewhere inside the native TF lib.

Some current observations on that:
Ignoring any waiting threads (`__GI___poll`, `__pthread_cond_timedwait`, `futex_wait_cancelable`, `nsync_mu_semaphore_p_with_deadline`, etc), we see e.g.`TensorExecutorTilingContext` via `BinaryOp`, and `CopyDeviceToHost`, via `IntraProcessRecvAsyncImpl`, in multiple logs, and then also `SplitVOpGPU`, and `ReduceFunctor`, and `BinaryOp`, and `CopyHostToDevice` and `LaunchConv2DOp`. So this is a bit unclear.

**Comment by vieting** [comment]
I'm running a training that showed this extreme slowdown with TF 2.14. The slowdown occurs slightly later (starting from epoch 4, step 3862 instead of epoch 4, step 1853 with TF 2.8), but it still occurs and is significant. Currently, the step durations are at >100 sec/step. So if it is a memory leak on TF side, it's at least not fixed in the latest version.

**Comment by vieting** [comment]
I created a script to reproduce the issue with stand-alone TF code. Indeed, with a convolution on the waveform followed by pooling and a linear layer with CE loss on random data, the memory leak can be reproduced. When removing the convolution, there is no leak. Also, if the sequence lengths are short, I did not observe a leak.

![image](https://github.com/rwth-i6/returnn/assets/45091115/fb709883-86c2-44e5-953c-b12b11f772b7)

The script to reproduce the leak looks like this:

<details>

```
#!/usr/bin/env python3

"""
https://github.com/rwth-i6/returnn/issues/1450
"""

import inspect
import sys
import numpy
import tensorflow as tf
from watch_memory import watch_memory
watch_memory()
from returnn.tf.updater import NadamOptimizer

tf1 = tf.compat.v1

tf1.disable_eager_execution()
tf1.disable_control_flow_v2()


n_feat = 1
subsampl_factor = 80
n_hidden = 30
filter_size = 5
n_out = 20
n_steps = 100
n_epochs = sys.maxsize


def model(x):
    filters = tf.Variable(tf.random.normal((filter_size, n_feat, n_hidden), stddev=0.01))
    x = tf1.nn.convolution(
        x,
        data_format=None,
        filter=filters,
        padding="VALID",
        strides=[1],
        dilation_rate=[1],
    )
    x = tf1.nn.pool(
        x,
        window_shape=(subsampl_factor,),
        pooling_type="AVG",
        padding="VALID",
        dilation_rate=[1],
        strides=(subsampl_factor,),
        data_format=None,
    )
    weights = tf.Variable(tf.random.normal((n_hidden, n_out), stddev=0.01))
    x = tf.matmul(x, weights)
    return x


def run_epoch(model_f):
    print("Setup TF graph")
    graph = tf1.Graph()
    with graph.as_default():
        print("Setup TF session")
        session = tf1.Session(graph=graph)
        with session.as_default():
            print("Setup model and computation")
            x = tf1.placeholder(tf.float32, shape=(None, None, n_feat))
            targets = tf1.placeholder(tf.int32, shape=(None, None))
            y = model_f(x)
            targets_cropped = targets[:, :tf.shape(y)[1]]
            loss = tf.reduce_sum(tf.nn.sparse_softmax_cross_entropy_with_logits(logits=y, labels=targets_cropped))
            opt = NadamOptimizer(learning_rate=0.01).minimize(loss)
            print("Initialize variables")
            session.run(tf1.global_variables_initializer())
            print("Run epoch")
            for step in range(n_steps):
                n_time = numpy.random.randint(9000, 120000)
                n_batch = 400000 // n_time  # numpy.random.randint(1, 10)
                feed_dict = {
                    x: numpy.random.normal(0.0, 1.0, (n_batch, n_time, n_feat)).astype(numpy.float32),
                    targets: numpy.random.randint(n_out, size=(n_batch, n_time // subsampl_factor)).astype(numpy.int32)
                }
                _, loss_v = session.run((opt, loss), feed_dict=feed_dict)
            print("Last loss:", loss_v / (n_batch * n_time))


def main():
    print("Code of executed script:")
    print(inspect.getsource(sys.modules[__name__]))
    for epoch in range(n_epochs):
        print(f"*** Epoch {epoch}", flush=True)
        run_epoch(model)


if __name__ == '__main__':
    main()

```

</details>

**Comment by vieting** [comment]
The script can even be simplified quite a lot and still reproduce the memory growth. Even when just running the convolution (no other operations or loss computations), the memory grows linearly. Also after changing to TF 2.14, the behavior is the same.

```
import sys
import numpy
import tensorflow as tf
from watch_memory import watch_memory
watch_memory()


n_feat = 1
n_hidden = 30
filter_size = 5
n_steps = 100
n_epochs = sys.maxsize


def run_epoch(filters):
    print("Run epoch")
    for step in range(n_steps):
        n_time = numpy.random.randint(9000, 120000)
        n_batch = 400000 // n_time
        x = tf.random.normal((n_batch, n_time, n_feat))
        y = tf.nn.convolution(
            x,
            filters=filters,
            data_format=None,
            padding="VALID",
            strides=[1],
            dilations=[1],
        )


def main():
    filters = tf.Variable(tf.random.normal((filter_size, n_feat, n_hidden), stddev=0.01))
    for epoch in range(n_epochs):
        print(f"*** Epoch {epoch}", flush=True)
        run_epoch(filters)


if __name__ == '__main__':
    main()
```

**Comment by albertz** [comment]
Ah very interesting. Initially, you said, you cannot reproduce it? So what change was relevant now? Having the larger batch size?

**Comment by albertz** [comment]
Can you report that in the TensorFlow GitHub issues and link it here?

**Comment by albertz** [comment]
If you take the same script but replace TF by PyTorch, how does that behave?

**Comment by albertz** [comment]
If you play around with some other things, e.g. `n_feat` or `n_hidden`, does the mem leak still occur?
